src/XR/features/WebXRMeshDetector.ts

```diff
@@ -0,0 +1,238 @@
+                    const newMesh: Partial<IWebXRMesh> = {
+                        id: meshIdProvider++,
+                        xrMesh: xrMesh,
+                    };
+                    const mesh = this._updateMeshWithXRMesh(xrMesh, newMesh, frame);
+                    this._detectedMeshes.push(mesh);
+                    this.onMeshAddedObservable.notifyObservers(mesh);
+                } else {
+                    // updated?
+                    if (xrMesh.lastChangedTime === this._xrSessionManager.currentTimestamp) {
+                        let index = this._findIndexInMeshArray(xrMesh);
+                        const mesh = this._detectedMeshes[index];
+                        this._updateMeshWithXRMesh(xrMesh, mesh, frame);
+                        this.onMeshUpdatedObservable.notifyObservers(mesh);
+                    }
+                }
+            });
+            this._lastDetectedSet = detectedMeshes;
+        }
+    }
+
+    private _initMeshDetector(session: XRSession) {
+        if (!!session.trySetMeshDetectorEnabled) {
+            session.trySetMeshDetectorEnabled(true);
+        }
+
+        if (!!this._options.preferredDetectorOptions &&
+            !!session.trySetPreferredMeshDetectorOptions) {
+            session.trySetPreferredMeshDetectorOptions(this._options.preferredDetectorOptions);
+        }
+    }
+
+    private _updateMeshWithXRMesh(xrMesh: XRMesh, mesh: Partial<IWebXRMesh>, xrFrame: XRFrame): IWebXRMesh {
+        mesh.xrMesh = xrMesh;
+        if (!this._xrSessionManager.scene.useRightHandedSystem) {
+            mesh.positions = new Float32Array(xrMesh.positions.length);
+            for (let i = 0; i < xrMesh.positions.length; i += 3) {
+                mesh.positions[i] = xrMesh.positions[i];
+                mesh.positions[i + 1] = xrMesh.positions[i + 1];
+                mesh.positions[i + 2] = -1 * xrMesh.positions[i + 2];
+            }
+
+            mesh.indices = new Uint32Array(xrMesh.indices.length);
+            for (let i = 0; i < xrMesh.indices.length; i += 3) {
+                mesh.indices[i] = xrMesh.indices[i];
+                mesh.indices[i + 1] = xrMesh.indices[i + 2];
+                mesh.indices[i + 2] = xrMesh.indices[i + 1];
+            }
+
+            if (!!xrMesh.normals) {
+                mesh.normals = new Float32Array(xrMesh.normals.length);
+                for (let i = 0; i < xrMesh.normals.length; i += 3) {
+                    mesh.normals[i] = xrMesh.normals[i];
+                    mesh.normals[i + 1] = xrMesh.normals[i + 1];
+                    mesh.normals[i + 2] = -1 * xrMesh.normals[i + 2];
+                }
+            }
+        }
+        else {
+            mesh.positions = xrMesh.positions;
+            mesh.indices = xrMesh.indices;
+            mesh.normals = xrMesh.normals;
+        }
+
+        mesh.transformationMatrix = this._options.worldParentNode?.getWorldMatrix() ?? Matrix.Identity();
+        return <IWebXRMesh>mesh;

```

Comment:
expanding the discussion earlier about the mesh vs. geometry - this object is actually not a mesh. A mesh should extend BABYLON.AbstractMesh (or TransformNode), otherwise it will be confusing for the developer.

I have internally discussed (with myself), whether or not the plane detector should construct a PlaneMesh on each frame, or let the user do it. As I assumed people will use those polygons in different ways, i have decided not to do that. But in the case of this one, generating (and constantly updating) a Babylon VertextBuffer with this data (or maybe BABYLON.Geometry?) would make a lot of sense. You can use the geometry to generate a mesh, you can keep a reference to it and constantly update it. 

One could argue that the plane detector could do the same, and I would probably agree with that :-) The issue with the plane detector is that we have a polygon builder that takes points and space on a single plane and extrude them, which is a bit different that generating a reusable geometry (as it builds a mesh directly).